File: longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        // Check if the input vector is empty.
        if(nums.size() == 0)
            return 0;

        // Create a vector to store the elements of the LIS in a dynamic manner.
        vector<int> lis;
        
        // Initialize the LIS vector with the first element of the input vector.
        lis.push_back(nums[0]);

        // Iterate through the remaining elements of the input vector to build the LIS.
        for(int i = 1; i < nums.size(); ++i){
            // If the current element is greater than the last element in the LIS,
            // append it to the LIS as it contributes to the increasing subsequence.
            if(nums[i] > lis.back()){
                lis.push_back(nums[i]);
            }
            // If the current element is not greater, find the position where it can be inserted
            // using binary search and update the LIS accordingly.
            else{
                int idx = lower_bound(lis.begin(), lis.end(), nums[i]) - lis.begin();
                lis[idx] = nums[i];
            }
        }

        // The size of the LIS vector represents the length of the Longest Increasing Subsequence.
        return lis.size();
    }
};

File: longest-increasing-subsequence/longest-increasing-subsequence.java
import java.util.*;
class Solution {
    public int lengthOfLIS(int[] nums) {
        // Check if the input array is empty
        if(nums.length == 0)
            return 0;

        //Create an array to store the element of LIS is a dynamic manner.
        int[] lis = new int[nums.length];
        Arrays.fill(lis, Integer.MAX_VALUE);

        // Intetialize the LIS array with the first element of the input array.
        lis[0] = nums[0];

        // Variable to keep track of the length of the LIS
        int len = 1;

        // Iterate through the remaining elements of the input array to build LIS.
        for(int i=1; i<nums.length; ++i){
            // If the current element is greater than the last element in the LIS
            // Append it to the LIS as it contributes to the increasing subsequence
            if(nums[i] > lis[len-1]){
                lis[len++] = nums[i];
            }
            // If the current element is not greater, find the position where it can be inserted.
            // Using binary search and update the LIS accordingly
            else{
                int idx = Arrays.binarySearch(lis, 0, len, nums[i]);
                if(idx<0){
                    idx = -idx - 1;
                }
                lis[idx] = nums[i];
            }
        }
        // The value of 'len' represents the length of the Longest Increasing Subsequence.
        return len;
    }
}